linabutler/gecko-dev
# Skeleton of a Rust login manager. ☠️

Let's start on our logins prototype! 🔐 

(Copy-pasting from the commit message...)

This PR adds all the boilerplate for a Rust login manager that
wraps our logins component. There are a few things to note:

* We're gating the Rust component behind the `MOZ_RUST_LOGINS` config
  flag (enabled unconditionally now), so that we can eventually ship it
  side-by-side with the existing JS implementation, and maybe only
  build it on Nightly until we can replace the latter.
* Rust XPCOM doesn't support `jsval`s at all (unfortunately, this means
  no promises; we have to use callbacks), so we've split those methods
  out into a `nsILoginManagerBase` interface, which we implement in
  Rust, and added a `RustLoginManagerWrapper` that implements the
  three async methods.
* `nsILoginManager` is synchronous. For a prototype, this is fine, but
  we can't ship it. If we want to implement async calls in Rust, we'll
  probably need to add a new `nsILoginManagerAsync` interface, with
  callback signatures like
  `void addLogins(in Array<nsILoginInfo> logins,
  in nsIAddLoginCallback)` instead of
  `jsval addLogins(in jsval aLogins)`.

Rust changes require a full `mach build`, which takes a while (about
3-4 minutes on my machine). To speed up development, you can run
`cargo check` on just the `passowrdmgr` crate, like this:

    env MOZ_TOPOBJDIR=/path/to/objdir cargo check -p passwordmgr

(`cargo fmt` works the same way).

To test out the new skeleton, run:

    ./mach xpcshell-test toolkit/components/passwordmgr/test/unit/test_rust_logins.js

You'll see "Adding login in Rust", and then the test will exit with
`NS_ERROR_NOT_IMPLEMENTED`.
